File: frontend/src/utils/device.js
const DEVICE_LABELS = {
  mobile: '移动端',
  tablet: '平板端',
  desktop: '桌面端',
}

function detectPlatform(userAgent = '') {
  const ua = userAgent.toLowerCase()
  if (ua.includes('iphone') || ua.includes('ipad') || ua.includes('ipod')) return 'ios'
  if (ua.includes('android')) return 'android'
  if (ua.includes('windows')) return 'windows'
  if (ua.includes('mac os x') || ua.includes('macintosh')) return 'macos'
  if (ua.includes('linux')) return 'linux'
  return 'unknown'
}

function detectBrowser(userAgent = '') {
  const ua = userAgent.toLowerCase()
  if (ua.includes('edg/')) return 'edge'
  if (ua.includes('chrome/') && !ua.includes('edg/')) return 'chrome'
  if (ua.includes('firefox/')) return 'firefox'
  if (ua.includes('safari/') && !ua.includes('chrome/')) return 'safari'
  return 'unknown'
}

function detectDeviceType({ width = 1440, userAgent = '', isTouch = false }) {
  const ua = userAgent.toLowerCase()
  const isTabletUa =
    ua.includes('ipad')
    || ua.includes('tablet')
    || ua.includes('playbook')
    || ua.includes('kindle')
    || ua.includes('silk')
    || (ua.includes('android') && !ua.includes('mobile'))
  const isMobileUa =
    ua.includes('iphone')
    || ua.includes('ipod')
    || ua.includes('windows phone')
    || ua.includes('opera mini')
    || (ua.includes('android') && ua.includes('mobile'))

  if (width <= 768) return 'mobile'
  if (width <= 1180 && (isTouch || isTabletUa || isMobileUa)) return 'tablet'
  if (isTabletUa) return 'tablet'
  if (isMobileUa) return 'mobile'
  if (width <= 1180) return 'tablet'
  return 'desktop'
}

export function getClientDeviceInfo() {
  if (typeof window === 'undefined') {
    return {
      deviceType: 'desktop',
      deviceLabel: DEVICE_LABELS.desktop,
      viewportWidth: 1440,
      viewportHeight: 900,
      orientation: 'landscape',
      platform: 'unknown',
      browser: 'unknown',
      isTouch: false,
      isMobileLayout: false,
      isTabletLayout: false,
      isCompactLayout: false,
    }
  }

  const userAgent = window.navigator.userAgent || ''
  const viewportWidth = window.innerWidth || document.documentElement.clientWidth || 1440
  const viewportHeight = window.innerHeight || document.documentElement.clientHeight || 900
  const hasCoarsePointer =
    typeof window.matchMedia === 'function'
      ? window.matchMedia('(pointer: coarse)').matches
      : false
  const isTouch = Boolean(window.navigator.maxTouchPoints > 0 || hasCoarsePointer)
  const deviceType = detectDeviceType({ width: viewportWidth, userAgent, isTouch })
  const isMobileLayout = viewportWidth < 992
  const isTabletLayout = !isMobileLayout && viewportWidth < 1280

  return {
    deviceType,
    deviceLabel: DEVICE_LABELS[deviceType] || DEVICE_LABELS.desktop,
    viewportWidth,
    viewportHeight,
    orientation: viewportWidth >= viewportHeight ? 'landscape' : 'portrait',
    platform: detectPlatform(userAgent),
    browser: detectBrowser(userAgent),
    isTouch,
    isMobileLayout,
    isTabletLayout,
    isCompactLayout: viewportWidth < 1280,
  }
}

export function getClientDeviceHeaders() {
  const info = getClientDeviceInfo()

  return {
    'X-CY-Device-Type': info.deviceType,
    'X-CY-Viewport-Width': String(info.viewportWidth),
    'X-CY-Viewport-Height': String(info.viewportHeight),
    'X-CY-Touch-Capable': info.isTouch ? 'true' : 'false',
    'X-CY-Platform': info.platform,
  }
}

export default getClientDeviceInfo

File: backend/services/device.py
"""客户端设备识别工具。"""

from __future__ import annotations

from typing import Mapping, Optional

VALID_DEVICE_TYPES = {"mobile", "tablet", "desktop"}


def _normalize_header_value(value: Optional[str]) -> Optional[str]:
    if value is None:
        return None
    normalized = value.strip().lower()
    return normalized or None


def _parse_int(value: Optional[str]) -> Optional[int]:
    if value is None:
        return None
    try:
        return int(value)
    except (TypeError, ValueError):
        return None


def detect_device_context(headers: Mapping[str, str]) -> dict:
    lowered_headers = {str(key).lower(): value for key, value in headers.items()}
    user_agent = lowered_headers.get("user-agent", "") or ""
    normalized_ua = user_agent.lower()

    requested_device_type = _normalize_header_value(lowered_headers.get("x-cy-device-type"))
    viewport_width = _parse_int(lowered_headers.get("x-cy-viewport-width"))
    viewport_height = _parse_int(lowered_headers.get("x-cy-viewport-height"))
    touch_capable = _normalize_header_value(lowered_headers.get("x-cy-touch-capable")) == "true"

    is_tablet_ua = any(
        token in normalized_ua
        for token in ("ipad", "tablet", "playbook", "kindle", "silk")
    ) or ("android" in normalized_ua and "mobile" not in normalized_ua)
    is_mobile_ua = any(
        token in normalized_ua
        for token in ("iphone", "ipod", "windows phone", "opera mini")
    ) or ("android" in normalized_ua and "mobile" in normalized_ua)

    if requested_device_type in VALID_DEVICE_TYPES:
        device_type = requested_device_type
        detected_from = "client-header"
    elif is_tablet_ua:
        device_type = "tablet"
        detected_from = "user-agent"
    elif is_mobile_ua:
        device_type = "mobile"
        detected_from = "user-agent"
    elif viewport_width is not None:
        if viewport_width <= 768:
            device_type = "mobile"
        elif viewport_width <= 1180:
            device_type = "tablet"
        else:
            device_type = "desktop"
        detected_from = "viewport"
    else:
        device_type = "desktop"
        detected_from = "fallback"

    if "iphone" in normalized_ua or "ipad" in normalized_ua or "ipod" in normalized_ua:
        platform = "ios"
    elif "android" in normalized_ua:
        platform = "android"
    elif "windows" in normalized_ua:
        platform = "windows"
    elif "mac os x" in normalized_ua or "macintosh" in normalized_ua:
        platform = "macos"
    elif "linux" in normalized_ua:
        platform = "linux"
    else:
        platform = "unknown"

    if "edg/" in normalized_ua:
        browser = "edge"
    elif "chrome/" in normalized_ua and "edg/" not in normalized_ua:
        browser = "chrome"
    elif "firefox/" in normalized_ua:
        browser = "firefox"
    elif "safari/" in normalized_ua and "chrome/" not in normalized_ua:
        browser = "safari"
    else:
        browser = "unknown"

    return {
        "device_type": device_type,
        "platform": platform,
        "browser": browser,
        "viewport_width": viewport_width,
        "viewport_height": viewport_height,
        "is_touch": touch_capable or is_mobile_ua or is_tablet_ua,
        "detected_from": detected_from,
        "user_agent": user_agent,
    }
